Sources/FoundationEssentials/String/StringProtocol+Essentials.swift

```diff
@@ -21,6 +21,28 @@ extension String {
         switch encoding {
         case .utf8:
             return Data(self.utf8)
+        case .ascii, .nonLossyASCII:
+            if allowLossyConversion {
+                let lossyReplacement = (encoding == .ascii) ? 0xFF : UInt8(ascii: "?")
+                return Data(capacity: self.utf8.count) {
+                    for scalar in self.unicodeScalars {
+                        if scalar.isASCII {
+                            $0.append(fromContentsOf: scalar.utf8)
+                        } else {
+                            $0.appendElement(lossyReplacement)
+                        }
+                    }
+                }
+            } else {
+                var data = Data(count: self.utf8.count)
+                let allASCII = data.withUnsafeMutableBytes {
+                    $0.withMemoryRebound(to: UInt8.self) { buffer in
+                        buffer.initializeAll(fromContentsOf: self.utf8)
+                        return _allASCII(UnsafeBufferPointer(buffer))
```

Comment:
I'm not sure about this, but since we have to do this check every time - does it make sense to do it first? Then, if it fails, we do no allocation or copy.

Reply:
I can move this check beforehand if the `String` has a contiguous buffer available to perform the check on. In the case where the `String` doesn't have a contiguous storage to perform this check (i.e. it's a lazily bridged string from ObjC) - it's probably best to perform the check afterwards since performing the check beforehand will need to transcode everything anyways

Reply:
Added the early check (when we have a contiguous buffer) in e45250b

Reply:
Makes sense, thanks.